cranelift/isle/isle/src/parser.rs

```diff
@@ -56,6 +56,17 @@ impl<'a> Parser<'a> {
         }
     }
 
+    fn eat<F: Fn(&Token) -> bool>(&mut self, f: F) -> Result<Option<Token>> {
+        if let Some(&(_pos, ref peek)) = self.lexer.peek() {
+            if !f(peek) {
+                return Ok(None);
+            }
+            Ok(Some(self.lexer.next()?.unwrap().1))
+        } else {
+            Err(self.error(self.lexer.pos(), "Unexpected EOF".to_string()))
+        }
+    }
```

Comment:
I would have expected `eat` to return `Some(token)` if there is a next token and it matches the predicate, and on the other hand, return `None` if there's no more tokens or the next token doesn't match. I'm not convinced that returning a `Result` in order to distinguish the EOF case makes sense here.

It doesn't exactly matter since everywhere this is used we expect more tokens to follow anyway, so it's okay to detect the EOF here, but I think this is simpler to use and its behavior is easier to explain if it simply returns `None` at EOF.

Reply:
Changed to return `Ok(None)` on EOF.